code/modules/mob/living/bot/bot.dm

```diff
@@ -400,19 +403,54 @@
+	if(A == null || B == null)
+		return 1
 
-	if(A == null || B == null) return 1
 	var/adir = get_dir(A,B)
 	var/rdir = get_dir(B,A)
-	if((adir & (NORTH|SOUTH)) && (adir & (EAST|WEST)))	//	diagonal
-		var/iStep = get_step(A,adir&(NORTH|SOUTH))
-		if(!LinkBlockedWithAccess(A,iStep, ID) && !LinkBlockedWithAccess(iStep,B,ID))
-			return 0
-
-		var/pStep = get_step(A,adir&(EAST|WEST))
-		if(!LinkBlockedWithAccess(A,pStep,ID) && !LinkBlockedWithAccess(pStep,B,ID))
-			return 0
-		return 1
+
+	if((adir & (NORTH|SOUTH)) && (adir & (EAST|WEST)))	//diagonal
+		if(DirBlockedWithAccess(A,adir, ID) || DirBlockedWithAccess(B,rdir, ID))	//Boarder blockers use bit values to check dirs, so it's safe to pass diagonal dirs here.
+			return 1
+
+		for(var/obj/O in B)
+			if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
+				return 1
+
+		//This mess is mainly to combat getting stuck in glass windows due to weird diagonal step_towards behaviour. If you're going northeast for example: if there's a window facing east on the tile above, there MUST be another one facing south or the east facing window will block the move, regardless if the tile to the right of the starting point is completely clear.
+
+		var/turf/yStep = get_step(A,adir&(NORTH|SOUTH))	//Vertical tile next to our start tile and end tile
+		var/turf/xStep = get_step(A,adir&(EAST|WEST))	//Horizontal tile next to our start tile and end tile
+
+		if(DirBlockedWithAccess(yStep,rdir&(NORTH|SOUTH),ID))	//One 'path' is blocked. The other path MUST be completely free
+			if(DirBlockedWithAccess(xStep,rdir&(EAST|WEST),ID) || DirBlockedWithAccess(xStep,adir&(NORTH|SOUTH),ID) || xStep.density)
+				return 1
+			else
+				for(var/obj/O in xStep)	//Check for dense objects in the other path
+					if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
+						return 1
+				return 0
+		else if(DirBlockedWithAccess(yStep,adir&(EAST|WEST),ID))	//Adjacent tile exiting dir is blocked but entering isn't. Mobs will get stuck here, even if the other path is free!
+			return 1
+
+		if(DirBlockedWithAccess(xStep,rdir&(EAST|WEST),ID))	//Same checks, but for the other path
+			if(yStep.density)	//We've already confirmed there's no boarder blockers in the other path, we just need a simple density check this time.
+				return 1
+			else
+				for(var/obj/O in yStep)
+					if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
+						return 1
+				return 0
+		else if(DirBlockedWithAccess(xStep,adir&(NORTH|SOUTH),ID))
+			return 1
+
+		for(var/obj/O in yStep)	//Both paths free? Check them for objects. So long as one isn't blocked, we can move
+			if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
+				for(var/obj/OB in xStep)
+					if(OB.density && !istype(OB, /obj/machinery/door) && !(OB.atom_flags & ATOM_FLAG_CHECKS_BORDER))
+						return 1
+				return 0
```

Comment:
I can't quite follow why the two for-loops are nested here. If it's a cardinal move, X/Y steps shouldn't influence each other.  Either we move horizontally and don't care about the vertical blockers, or we move vertically and we don't care about the horizontal ones.

If this is a diagonal move, surely we'd only need to check a single turf in a single loop as xStep & yStep refer to the same turf, right? Or is the idea that if we move diagonally, we need to have at least one valid step in both cardinal directions to avoid phasing through corners?

In either case, it doesn't seem like we ever need to compare O and OB here, so it probably should be two separate for-loops over yStep & xStep. This would reduce the complexity of this logic from `O(N^2)` to `O(2N) == O(N)` (or less abstractly - stop this proc from eating the server hardware raw just because someone put a big pile of items on both tiles).

Reply:
This is only triggered in the case of diagonal moves (within the if statement for diagonals on L412)
`xStep` and `yStep` are not the same turf, they're the in-between adjacent turfs on each axis, as byond diagonal movement considers a diagonal move to mean passing through 4 turfs, not 2. This is here to validate that at least one of the steps is clear of objects so the move can complete.

Although they are _technically_ nested, in this case the second `for` is acting more like a `break`. The idea is _either_ `xStep` or `yStep` must be clear of dense objects; if they both have dense objects, the diagonal move will fail.

If there's no dense objects in `yStep`, the second nested `for` loop on L449 is never triggered (`xStep` is never checked). If there is a dense object, the first loop would pause while the second loop checks for blockers in `xStep`. Whether there's blockers or not, that second loop will **always** call return (L451+L452), so the first loop is never ran again, again kinda working like a `break` would. The whole thing works like it's a giant AND check (if loop one is true, check the other loop. If that loop is also true, return 1, else 0)

The idea was to save an additional variable and check, and functions in the same way as below:
```dm
var/blocked = FALSE
for(var/obj/O in yStep)
    if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
        blocked = TRUE
        break
if(blocked)
    for(var/obj/O in xStep)
        if(O.density && !istype(O, /obj/machinery/door) && !(O.atom_flags & ATOM_FLAG_CHECKS_BORDER))
            return 1
return 0
```

I can understand it doesn't look as obvious as  the above, however, I thought it was a more elegant way of handling it. Happy to switch it over though it readability is a concern!

Reply:
Ah, yes, I see now, the fact that the second for-loop is gated behind an if and there's ALWAYS a return in that if means that the worst-case performance is still O(n+m), so that's fine.

Still, I'd prefer option 2 for clarity; variables are _cheap_

Reply:
Not a problem! Switched to the above example for readability.